zcm/julia/ZCM.jl

```diff
@@ -15,320 +59,276 @@ end
+    data       ::Ptr{UInt8}
+    data_size  ::UInt32
 end
 
 type EventLog
 end
 
 type EventLogEvent
-    eventnum  ::Int64;
-    timestamp ::Int64;
-    chanlen   ::Int32;
-    datalen   ::Int32;
-    channel   ::Ptr{UInt8};
-    data      ::Ptr{UInt8};
+    eventnum  ::Int64
+    timestamp ::Int64
+    chanlen   ::Int32
+    datalen   ::Int32
+    channel   ::Ptr{UInt8}
+    data      ::Ptr{UInt8}
 end
 
 end
 
-# RRR: make variable naming consistent
-# RRR: improve orgainzation and make header comments more visible
-# Exported Objects and Methods
-type RecvBuf
-    rbuf  ::Ptr{Native.RecvBuf};
-    utime ::Int64;
-    data  ::Array{UInt8};
-
-    function RecvBuf(rbuf::Ptr{Native.RecvBuf})
-        instance  = new();
-        instance.rbuf = rbuf;
+using .Native: RecvBuf
 
-        if (rbuf != C_NULL)
-            loadedRbuf = unsafe_load(rbuf);
-
-            instance.utime = loadedRbuf.recv_utime;
+"""
+The SubscriptionOptions type contains all of the information necessary to
+decode and handle a message on the Julia side. This is the structure which is
+passed in to zcm_subscribe, and it is thus also passed to the handler_wrapper()
+cfunction.
+"""
+immutable SubscriptionOptions{T <: AbstractZCMType, F}
+    msgtype::Type{T}
+    handler::F
+end
 
-            if (loadedRbuf.data != C_NULL)
-                instance.data = unsafe_wrap(Array, loadedRbuf.data, loadedRbuf.data_size);
-            else
-                instance.data = [];
-            end
-        end
+"""
+The Subscription type contains the SubscriptionOptions and also all of the
+various C pointers to the libuv handlers.
+"""
+immutable Subscription{T <: SubscriptionOptions}
+    options::T
+    c_handler::Ptr{Void}
+    uv_wrapper::Ptr{Native.UvSub}
+    uv_handler::Ptr{Void}
+    native_sub::Ptr{Native.Sub}
+end
 
-        # destruction of the underlying Native.RecvBuf pointer is handled by the C lib internally
+type Zcm
+    zcm::Ptr{Native.Zcm}
+    subscriptions::Vector{Subscription}
+
+    function Zcm(url::AbstractString = "")
+        pointer = ccall(("zcm_create", "libzcm"), Ptr{Native.Zcm}, (Cstring,), url);
+        instance = new(pointer, Subscription[])
+        finalizer(instance, close)
+        return instance
+    end
+end
 
-        return instance;
+function close(zcm::Zcm)
+    if zcm.zcm != C_NULL
+        ccall(("zcm_destroy", "libzcm"), Void,
+              (Ptr{Native.Zcm},), zcm)
+        zcm.zcm = C_NULL
     end
 end
-export RecvBuf;
 
-type __zcm_handler
-    handler::Any;
-    usr::Any;
+# Defines the conversion when we pass this to a C function expecting a pointer
+unsafe_convert(::Type{Ptr{Native.Zcm}}, zcm::Zcm) = zcm.zcm
 
-    # handler must have the function or functor prototype :
-    #   (rbuf::RecvBuf, channel::String, msg::MsgType, usr)
-    __zcm_handler(handler, usr) = new(handler, usr);
+function good(zcm::Zcm)
+    (zcm.zcm != C_NULL) && (errno(zcm) == 0)
```

Comment:
Assume you meant to return here?

Reply:
the result of the last statement in a function is always returned (that's also why the short form `f(x) = y` works)